pkg/utils/utils.go

```diff
@@ -291,6 +291,19 @@ func GetVFIODeviceFile(dev string) (devFile string, err error) {
 		return
 	}
 
+	// Get a file path to the iommu group name
+	namePath := filepath.Join(linkName, "name")
+	// Read the iommu group name
+	// The name file will not exist on baremetal
+	vfioName, errName := ioutil.ReadFile(namePath)
+	if errName == nil {
+		vName := strings.TrimSpace(string(vfioName))
+
+		// if the iommu group name == vfio-noiommu then we are in a VM, adjust path to vfio device
+		if vName == "vfio-noiommu" {
+			linkName = filepath.Join(filepath.Dir(linkName), "noiommu-"+filepath.Base(linkName))
```

Comment:
so the symbol link of iommuDir for no-iommu vfio device is the same as iommu vfio device, that's why we need to add `noiommu-` prefix here. Just curious the symbol link of iommuDir doesn't contain the actual device spec.

Reply:
The directory does contain a device symlink.  i.e.

sh-4.2# ls -l /sys/kernel/iommu_groups/0                        
total 0
drwxr-xr-x. 2 root root    0 Oct 13 14:14 devices
-r--r--r--. 1 root root 4096 Oct 12 17:21 name
-r--r--r--. 1 root root 4096 Oct 13 14:14 reserved_regions
-r--r--r--. 1 root root 4096 Oct 13 14:14 type

devices points to the device dir...

on virtual, the type and the name are present.

As a shortcut, I decided to key off of the presence and value of the name file.
The name file is present for the noiommu case and not present for the baremetal
case (in my testing).  In addition, the name file == vfio-noiommu.  I use this fact
as postive confirmation that we are running virtual.  If we are running on a virtual
platform with virtualized iommu, the name field should not be present and/or the
name will not be vfio-noiommu from looking at the kernel code.

The name file is references [here](https://www.kernel.org/doc/Documentation/ABI/testing/sysfs-kernel-iommu_groups).

Reply:
Aaron, thanks for the detailed explanation!

Reply:
I'm not very clear on this. What would be the deviceSpec for noiommu vfio devices?
Which device files do we need to mount in the container?
is this going to be always:
`/dev/vfio/vfio` and 
`/dev/vfio/0`
?

Reply:
There is no additional deviceSpec required for noiommu.  Part of additional code determines if the noiommu device exists.  If the device, noiommu, exists, then it is used.  The  /dev/vfio/noiommu-0 device in the VM is mounted as /dev/vfio/0 inside the container.

Reply:
So, if the /dev/vfio/0 in host with noiommu will be mounted as /dev/vfio/0 in container why do we need this change here?

Reply:
Sorry, I was unclear/wrong about deviceSpec...   

in the VM case in the following section of code (vfrioResource.go,59), (HostPath) vfioDev == /dev/vfio/noiommu-0.  
The containerPath in this case need to be /dev/vfio/0, hence the replace.

		devSpecs = append(devSpecs, &pluginapi.DeviceSpec{
			HostPath:      vfioDev,
			ContainerPath: strings.Replace(vfioDev, "noiommu-", "", 1),
			Permissions:   "mrw",
		})

Reply:
Isn't this being replaced because it was concatenated in L304? If we didn't add that in there there'd be no need to do the string replace here. I mean without adding "noiommu-" prefix to that iommu_group '0' we would still get the usual `/dev/vfio/0` as we would get in normal vfio case. 

What I am trying to understand is what do we need to do differently in "noiommu" cases?
Am I missing something? I don't have a VM instance to verify this properly.

Reply:
Inside the VM, when you load and bind vfio-pci, the device shows up as  /dev/vfio/noiommu-N. where N is 0,1,2, etc...   
Not /dev/vfio/0.  The current sriov-network-device-plugin code looks for /dev/vfio/0 on the host (which in this case is a VM).  /dev/vfio/0 does not exists.   The new code checks for the VM case by checking an optional *name* file that exists before the device has a special name instead of the default name.  If the *name* file is found and the name is vfio-noiommu, then the hostPath has to be /dev/vfio/noiommu-0, but the containerPath must remain /dev/vfio/0 to make things compatible across platforms.

Reply:
> Inside the VM, when you load and bind vfio-pci, the device shows up as /dev/vfio/noiommu-N. where N is 0,1,2, etc...

^This is what I wasn't aware of :) Thank you!